File: tools/ghidra/DecompileStringRefs.java
// Decompile functions that reference matching strings.
// Usage:
//   -postScript DecompileStringRefs.java pattern1 pattern2 ...

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Data;
import ghidra.program.model.listing.DataIterator;
import ghidra.program.model.listing.Function;
import ghidra.program.model.symbol.Reference;
import ghidra.util.task.ConsoleTaskMonitor;

public class DecompileStringRefs extends GhidraScript {

	private static class RefInfo {
		Address stringAddr;
		String text;
		Address fromAddr;

		RefInfo(Address stringAddr, String text, Address fromAddr) {
			this.stringAddr = stringAddr;
			this.text = text;
			this.fromAddr = fromAddr;
		}
	}

	@Override
	protected void run() throws Exception {
		String[] patterns = getScriptArgs();
		if (patterns == null || patterns.length == 0) {
			println("No patterns provided");
			return;
		}

		Map<Function, List<RefInfo>> funcs = new LinkedHashMap<>();
		int stringMatches = 0;

		DataIterator it = currentProgram.getListing().getDefinedData(true);
		while (it.hasNext()) {
			Data data = it.next();
			Object value = data.getValue();
			if (value == null) {
				continue;
			}

			String text = value.toString();
			if (text == null || text.isEmpty()) {
				continue;
			}

			boolean matched = false;
			for (String pat : patterns) {
				if (text.contains(pat)) {
					matched = true;
					break;
				}
			}
			if (!matched) {
				continue;
			}

			stringMatches++;
			Reference[] refs = getReferencesTo(data.getAddress());
			for (Reference ref : refs) {
				Function func = getFunctionContaining(ref.getFromAddress());
				if (func == null) {
					continue;
				}
				funcs.computeIfAbsent(func, k -> new ArrayList<>())
					.add(new RefInfo(data.getAddress(), text, ref.getFromAddress()));
			}
		}

		println("Matched " + stringMatches + " strings");
		println("Matched " + funcs.size() + " functions");

		DecompInterface ifc = new DecompInterface();
		ifc.openProgram(currentProgram);
		ConsoleTaskMonitor taskMonitor = new ConsoleTaskMonitor();

		for (Map.Entry<Function, List<RefInfo>> entry : funcs.entrySet()) {
			Function func = entry.getKey();
			println("=== FUNCTION " + func.getName() + " " + func.getEntryPoint() + " ===");
			for (RefInfo ref : entry.getValue()) {
				println("REF " + ref.fromAddr + " -> " + ref.stringAddr);
				println("STR " + ref.text);
			}

			DecompileResults res = ifc.decompileFunction(func, 60, taskMonitor);
			if (!res.decompileCompleted()) {
				println("DECOMPILATION FAILED: " + res.getErrorMessage());
				continue;
			}
			println(res.getDecompiledFunction().getC());
			println("");
		}

		ifc.dispose();
	}
}

File: tools/ghidra/decompile_string_refs.py
# Decompile functions that reference matching strings.
# Usage from analyzeHeadless:
#   -postScript decompile_string_refs.py pattern1 pattern2 ...

from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor


def iter_defined_data():
    listing = currentProgram.getListing()
    it = listing.getDefinedData(True)
    while it.hasNext():
        yield it.next()


def string_value(data):
    try:
        value = data.getValue()
        if value is None:
            return None
        text = str(value)
        if text:
            return text
    except:
        return None
    return None


def matching_data(patterns):
    matches = []
    for data in iter_defined_data():
        text = string_value(data)
        if not text:
            continue
        for pat in patterns:
            if pat in text:
                matches.append((data, text))
                break
    return matches


def funcs_for_matches(matches):
    funcs = {}
    for data, text in matches:
        refs = getReferencesTo(data.getAddress())
        for ref in refs:
            func = getFunctionContaining(ref.getFromAddress())
            if func is None:
                continue
            key = func.getEntryPoint().toString()
            if key not in funcs:
                funcs[key] = (func, [])
            funcs[key][1].append((data.getAddress(), text, ref.getFromAddress()))
    return funcs


def decompile_functions(func_map):
    ifc = DecompInterface()
    ifc.openProgram(currentProgram)
    monitor = ConsoleTaskMonitor()

    for key in sorted(func_map.keys()):
        func, refs = func_map[key]
        print("=== FUNCTION %s %s ===" % (func.getName(), func.getEntryPoint()))
        for saddr, text, fromaddr in refs:
            print("REF %s -> %s" % (fromaddr, saddr))
            print("STR %s" % text)
        res = ifc.decompileFunction(func, 60, monitor)
        if not res.decompileCompleted():
            print("DECOMPILATION FAILED: %s" % res.getErrorMessage())
            continue
        print(res.getDecompiledFunction().getC())
        print("")

    ifc.dispose()


if __name__ == "__main__":
    args = getScriptArgs()
    if not args:
        print("No patterns provided")
        exit(1)

    matches = matching_data(args)
    print("Matched %d strings" % len(matches))
    funcs = funcs_for_matches(matches)
    print("Matched %d functions" % len(funcs))
    decompile_functions(funcs)
